extensions/amp-document-recommendations/0.1/amp-document-recommendations.js

```diff
@@ -0,0 +1,177 @@
+      this.nextArticle_++;
+
+      Services.xhrFor(this.win)
+        .fetchDocument(next.ampUrl, {})
+        .then(
+          (doc) => {this.attachShadowDoc(doc);},
+          (error) => {});    
+    }
+  }
+
+  /** 
+   * @param {number} from
+   */
+  appendArticleLinks_(from) {
+    const doc = this.win.document;
+    let article = from;
+    const viewer = Services.viewerForDoc(this.getAmpDoc());
+
+    while (article < this.config_.recommendations.length &&
+        article - from < SEPARATPOR_RECOS) {
+      const next = this.config_.recommendations[article];
+      article++;
+
+      const articleHolder = doc.createElement('div');
+      articleHolder.classList.add('-reco-holder-article');
+      articleHolder.onclick = () => {
+        viewer.navigateTo(next.ampUrl, 'content-discovery');
+      }
+
+      const imageElement = doc.createElement('div');
+      imageElement.classList.add('-next-article-image', 'amp-document-recommendations-image');
+      imageElement.style.backgroundImage = `url(${next.image})`;
+      articleHolder.appendChild(imageElement);
+
+      const titleElement = doc.createElement('div');
+      titleElement.classList.add('-next-article-title', 'amp-document-recommendations-text');
+
+      titleElement.textContent = next.title;
+      articleHolder.appendChild(titleElement);
+
+      this.element.appendChild(articleHolder);
+
+      this.appendDivision_();
+    }
+  }
+
+  attachShadowDoc(doc, url) {
+    this.getVsync().mutate(() => {
+      const shadowRoot = this.win.document.createElement('div');
+      //shadowRoot.classList.add('amp-next-article-container');
+
+      try {
+        const amp = this.multidocManager_.attachShadowDoc(shadowRoot, doc, '', {});
+
+        // TODO(emarchiori): Update document title.
+
+        this.element.appendChild(shadowRoot);
+
+        if (this.nextArticle_ < this.config_.recommendations.length) {
+          this.appendNextArticle_();
+        }
+
+      } catch(e) {
+        this.handleLoadingError();
+      }
+    });
+  }
+
+  /** @override */
+  buildCallback() {
+    if (this.win.CONTENT_DISCOVERY !== this) {
+      return Promise.resolve();
+    }
+
+    this.multidocManager_ = new MultidocManager(
+      this.win,
+      Services.ampdocServiceFor(this.win),
+      Services.extensionsFor(this.win),
+      Services.timerFor(this.win));
+
+    const children = this.element.children;
+    user().assert(children.length == 1,
+        'The tag should contain exactly one <script> child.');
+    const scriptElement = children[0];
+    user().assert(
+        isJsonScriptTag(scriptElement),
+        'The amp-document-recommendations config should ' +
+        'be inside a <script> tag with type="application/json"');
+
+    const configJson = tryParseJson(scriptElement.textContent, error => {
+      throw user().createError('failed to parse content discovery script', error);
+    });
+
+    this.config_ = assertConfig(configJson);
+
+    return this.mutateElement(() => {
```

Comment:
Other components (amp-list, amp-date-picker) make XHR requests for their content from the layoutCallback - probably makes sense for this too?

Reply:
Done.